google-cloud-spanner/lib/google/cloud/spanner/pool.rb

```diff
@@ -166,20 +166,22 @@ def reset
 
         def close
           @mutex.synchronize do
+            @futures = nil
             @closed = true
           end
           @keepalive_task.shutdown
           # Unblock all waiting threads
           @resource.broadcast
           # Delete all sessions
           @mutex.synchronize do
-            @all_sessions.each { |s| future { s.release! } }
+            @futures = @all_sessions.map { |s| future { s.release! } }
             @all_sessions = []
             @session_queue = []
             @transaction_queue = []
           end
           # shutdown existing thread pool
           @thread_pool.shutdown
+          @futures.each(&:wait)
```

Comment:
The futures are already executing on the thread pool. so it would be simpler and potentially less error-prone to call `FixedThreadPool#wait_for_shutdown`. But we probably don't want to block on the thread pool shutdown when calling `#reset`, so we may need to take a slightly different approach here.